# NY 2019 Analysis

**Process:**
1. Upload NY 2019-2021 Data to `test` database
2. Companre new NY data with old data-- check for field differences
3. Fix or update field differences?
4. Upload NY 2019-2021 data to `ny` database


### 0. Set Up: Import Libraries & Connect to Mongo

In [1]:
import sys
sys.path.append("../../") 
import databasepopulation
import communicationwmongo as commu
import home_location as home
import analysis as a
import pymongo
import my_h3_functions as myh3

In C:\Users\Trang\anaconda3\envs\twitteranalysis\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The text.latex.preview rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\Users\Trang\anaconda3\envs\twitteranalysis\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The mathtext.fallback_to_cm rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\Users\Trang\anaconda3\envs\twitteranalysis\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: Support for setting the 'mathtext.fallback_to_cm' rcParam is deprecated since 3.3 and will be removed two minor releases later; use 'mathtext.fallback : 'cm' instead.
In C:\Users\Trang\anaconda3\envs\twitteranalysis\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The validate_bool_maybe_none function was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\Us

In [2]:
import importlib
importlib.reload(databasepopulation)
importlib.reload(home)
importlib.reload(a)
importlib.reload(myh3)

<module 'my_h3_functions' from '../..\\my_h3_functions.py'>

In [3]:
# Need to clean new data
import pandas as pd
from datetime import datetime, timezone
from dateutil.parser import parse

In [86]:
# What's the difference between this and following 2 cells?
#db=commu.connecttoLocaldb(database='test')

In [4]:
# Delete before pushing to Github
uri = 'mongodb://trang:trangUCber@3.14.72.122:27017/twitter?authSource=twitter&readPreference=primary&appname=MongoDB%20Compass%20Community&ssl=false'

In [5]:
client = pymongo.MongoClient(uri)
db = client['test']

In [5]:
# #test one
# db.test.insert_one({'test':True})
# db.test.find_one()

# ## Need to delete one in Sydney DB (accident)
# Output: {'_id': ObjectId('5f89e9bffadbd17469767bd6'), 'test': 'this is Jen testing'}

{'_id': ObjectId('5f89e9bffadbd17469767bd6'), 'test': 'this is Jen testing'}

## 1. Compare Differences

Notes: Many of fields in new data is not present inthe old data. Field names have periods in them, which will not work with Mongo (need to rename). Some field names are different than the old data (lat and lon). `created_at` column is not in epoch time like the original -- it's the date instead.



### Field Match with ny_2012:<br>
**id:** `id`<br>
**u_id:** `user.id` (?)<br>
**lon:** `coordinates.coordinates.coordinates.0`<br>
**lat:** `coordinates.coordinates.coordinates.1`<br>
**date:** ?<br>
**created_at:**`created_at`

### 1a. Read in old data to check fields.

In [33]:
ny12_file = "/Users/Trang/Box/Twitter data/North America/ny_2012.csv"

In [34]:
ny12 = pd.read_csv(ny12_file, nrows=5)
ny12.head()

Unnamed: 0                  id           u_id        lat        lon  \
0           0  228290876537970688  1.112537e-308  40.636800 -73.659319   
1           1  228290854564012032  1.112537e-308  40.778059 -73.954124   
2           2  228290916660674562  1.112537e-308  40.918340 -73.804538   
3           3  228290921916137474  1.112537e-308  40.923364 -74.146670   
4           4  228290919047262208  1.112537e-308  40.694654 -73.733507   

                        date     created_at  
0  2012-07-26 00:49:21+00:00  1343263761000  
1  2012-07-26 00:49:15+00:00  1343263755000  
2  2012-07-26 00:49:30+00:00  1343263770000  
3  2012-07-26 00:49:31+00:00  1343263771000  
4  2012-07-26 00:49:31+00:00  1343263771000

In [35]:
ny12.columns

Index(['Unnamed: 0', 'id', 'u_id', 'lat', 'lon', 'date', 'created_at'], dtype='object')

In [36]:
ny12.dtypes

Unnamed: 0      int64
id              int64
u_id          float64
lat           float64
lon           float64
date           object
created_at      int64
dtype: object

### 1b. Read in new file and get schema

In [76]:
ny19_file = "/Users/Trang/Box/Twitter data/North America/NY 2019-21/newyork/ny_2019_partial.csv"

In [77]:
ny19 = pd.read_csv(ny19_file)
ny19.head()

id                            date        lon        lat  \
0  1.253840e+18  Fri Apr 24 23:56:13 +0000 2020 -73.808799  40.688999   
1  1.253840e+18  Fri Apr 24 23:56:13 +0000 2020 -73.973101  40.608886   
2  1.253840e+18  Fri Apr 24 23:57:34 +0000 2020 -75.140060  40.023353   
3  1.253840e+18  Fri Apr 24 23:59:12 +0000 2020 -74.301162  40.491283   
4  1.248490e+18  Fri Apr 10 06:15:36 +0000 2020 -75.058488  40.035489   

           u_id  
0  5.070669e+07  
1  5.227294e+07  
2  9.689260e+17  
3  5.070669e+07  
4  3.601150e+08

In [61]:
len(ny19)

1048575

In [9]:
'''
Columns Before (Now on ny_2019_full.csv)
['id', 'created_at', 'truncated', 'text', 'is_quote_status',
       'in_reply_to_status_id', 'favorite_count', 'retweet_count', 'source',
       'coordinates.type', 'coordinates.coordinates.type',
       'coordinates.coordinates.coordinates.0',
       'coordinates.coordinates.coordinates.1', 'in_reply_to_user_id',
       'possibly_sensitive', 'lang', 'place.id', 'place.url',
       'place.place_type', 'place.name', 'place.full_name',
       'place.country_code', 'place.country', 'place.bounding_box.type',
       'place.bounding_box.coordinates.0.0.0',
       'place.bounding_box.coordinates.0.0.1',
       'place.bounding_box.coordinates.0.1.0',
       'place.bounding_box.coordinates.0.1.1',
       'place.bounding_box.coordinates.0.2.0',
       'place.bounding_box.coordinates.0.2.1',
       'place.bounding_box.coordinates.0.3.0',
       'place.bounding_box.coordinates.0.3.1', 'place.attributes',
       'filter_level', 'user.id', 'user.verified',
       'user.profile_image_url_https', 'user.followers_count', 'user.location',
       'user.utc_offset', 'user.statuses_count', 'user.description',
       'user.friends_count', 'user.screen_name', 'user.lang',
       'user.favourites_count', 'user.name', 'user.url', 'user.created_at',
       'user.time_zone', 'user.listed_count', 'hashtags', 'urls', 'media',
       'random_number', 'country', 'u_description']
'''

"\nColumns Before (Now on ny_2019_full.csv)\n['id', 'created_at', 'truncated', 'text', 'is_quote_status',\n       'in_reply_to_status_id', 'favorite_count', 'retweet_count', 'source',\n       'coordinates.type', 'coordinates.coordinates.type',\n       'coordinates.coordinates.coordinates.0',\n       'coordinates.coordinates.coordinates.1', 'in_reply_to_user_id',\n       'possibly_sensitive', 'lang', 'place.id', 'place.url',\n       'place.place_type', 'place.name', 'place.full_name',\n       'place.country_code', 'place.country', 'place.bounding_box.type',\n       'place.bounding_box.coordinates.0.0.0',\n       'place.bounding_box.coordinates.0.0.1',\n       'place.bounding_box.coordinates.0.1.0',\n       'place.bounding_box.coordinates.0.1.1',\n       'place.bounding_box.coordinates.0.2.0',\n       'place.bounding_box.coordinates.0.2.1',\n       'place.bounding_box.coordinates.0.3.0',\n       'place.bounding_box.coordinates.0.3.1', 'place.attributes',\n       'filter_level', 'user.id'

In [62]:
# Columns After: Renamed {'created_at':'date, 'coordinates.coordinates.coordinates.0: 'lon', 'coordinates.coordinates.coordinates.1': lat, 'user.id':u_id}
ny19.columns

Index(['id', 'date', 'lon', 'lat', 'u_id'], dtype='object')

In [63]:
ny19.dtypes

id      float64
date     object
lon     float64
lat     float64
u_id    float64
dtype: object

In [78]:
# Check date column to fix in next section

ny19.loc[0]['date']

'Fri Apr 24 23:56:13 +0000 2020'

### 2. Fix differences -- date and created_at fields

In [65]:
# Format of date string in new data
FORMAT = '%a %b %d %H:%M:%S %z %Y'

In [66]:
## Sample with one date (See Extra section for more details)
date_str = 'Fri Apr 24 23:56:13 +0000 2020'

date_obj = datetime.strptime(date_str, FORMAT)
date_obj_str = str(date_obj)

print('expected:', date_obj_str)
print('epoch time:', date_obj.replace(tzinfo=timezone.utc).timestamp() * 1000)

expected: 2020-04-24 23:56:13+00:00
epoch time: 1587772573000.0


### 2a. Convert `date` (previously `created_at`) column to datetime object

Question: Does it need to be converted into a string after making `created_at` column? Currently "Name: date, Length: 1048575, dtype: datetime64[ns, UTC]"

In [67]:
ny19.loc[:,'date'] =  pd.to_datetime(ny19['date'], format= FORMAT)
ny19.head()

id                      date        lon        lat          u_id
0  1.253840e+18 2020-04-24 23:56:13+00:00 -73.808799  40.688999  5.070669e+07
1  1.253840e+18 2020-04-24 23:56:13+00:00 -73.973101  40.608886  5.227294e+07
2  1.253840e+18 2020-04-24 23:57:34+00:00 -75.140060  40.023353  9.689260e+17
3  1.253840e+18 2020-04-24 23:59:12+00:00 -74.301162  40.491283  5.070669e+07
4  1.248490e+18 2020-04-10 06:15:36+00:00 -75.058488  40.035489  3.601150e+08

### 2b. Create Epoch Time `created_at` column using `date` column

In [18]:
ny19.loc[:,'created_at'] = ny19['date'].apply(lambda x: x.replace(tzinfo=timezone.utc).timestamp() * 1000)
ny19.head()

id                      date        lon        lat          u_id  \
0  1.253840e+18 2020-04-24 23:56:13+00:00 -73.808799  40.688999  5.070669e+07   
1  1.253840e+18 2020-04-24 23:56:13+00:00 -73.973101  40.608886  5.227294e+07   
2  1.253840e+18 2020-04-24 23:57:34+00:00 -75.140060  40.023353  9.689260e+17   
3  1.253840e+18 2020-04-24 23:59:12+00:00 -74.301162  40.491283  5.070669e+07   
4  1.248490e+18 2020-04-10 06:15:36+00:00 -75.058488  40.035489  3.601150e+08   

     created_at  
0  1.587773e+12  
1  1.587773e+12  
2  1.587773e+12  
3  1.587773e+12  
4  1.586499e+12

### 2b. Convert `date` column back to a string type before exporting


In [68]:
ny19.loc[:,'date'] =  ny19['date'].apply(str)
ny19['date']

0          2020-04-24 23:56:13+00:00
1          2020-04-24 23:56:13+00:00
2          2020-04-24 23:57:34+00:00
3          2020-04-24 23:59:12+00:00
4          2020-04-10 06:15:36+00:00
5          2020-04-10 06:16:36+00:00
6          2020-04-10 06:16:36+00:00
7          2020-04-10 06:16:36+00:00
8          2020-04-10 06:16:37+00:00
9          2020-04-28 03:19:21+00:00
10         2020-04-13 06:20:05+00:00
11         2020-04-02 12:35:45+00:00
12         2020-04-29 05:20:14+00:00
13         2020-04-04 15:25:21+00:00
14         2020-04-24 02:19:34+00:00
15         2020-04-02 12:37:13+00:00
16         2020-04-04 15:25:44+00:00
17         2020-04-15 04:57:33+00:00
18         2020-04-11 12:51:46+00:00
19         2020-04-11 12:51:47+00:00
20         2020-04-11 12:53:01+00:00
21         2020-04-01 12:48:45+00:00
22         2020-04-01 12:48:51+00:00
23         2020-04-01 12:49:41+00:00
24         2020-04-18 04:48:52+00:00
25         2020-04-29 05:23:09+00:00
26         2020-04-06 07:53:04+00:00
2

### 2c. Save file to populate tweets (`ny_2019.csv`)

In [74]:
PATH = '/Users/Trang/Box/Twitter data/North America/NY 2019-21/newyork/'

In [75]:
ny19.to_csv(PATH+'ny_2019.csv')

## 3. Upload NY '19-'21 Data Onto `test`

*06/15/21* -- **Error**: <br>
`InvalidDocument: key 'coordinates.type' must not contain '.'`

In [6]:
db

Database(MongoClient(host=['3.14.72.122:27017'], document_class=dict, tz_aware=False, connect=True, authsource='twitter', readpreference='primary', appname='MongoDB Compass Community', ssl=False), 'ny2019')

In [80]:
db.test

Collection(Database(MongoClient(host=['3.14.72.122:27017'], document_class=dict, tz_aware=False, connect=True, authsource='twitter', readpreference='primary', appname='MongoDB Compass Community', ssl=False), 'test'), 'test')

In [81]:
db.tweets

Collection(Database(MongoClient(host=['3.14.72.122:27017'], document_class=dict, tz_aware=False, connect=True, authsource='twitter', readpreference='primary', appname='MongoDB Compass Community', ssl=False), 'test'), 'tweets')

In [89]:
databasepopulation.populatetweets(db, path=PATH, cityprefix='ny', yearstart=2019, yearend=2019)

Now populating year  2019
process completed


## 4. Upload NY '19-'21 Data Onto `ny2019`

In [10]:
PATH = '/Users/Trang/Box/Twitter data/North America/NY 2019-21/newyork/'

In [12]:
client = pymongo.MongoClient(uri)
ny_db = client['ny2019']

In [13]:
ny_db

Database(MongoClient(host=['3.14.72.122:27017'], document_class=dict, tz_aware=False, connect=True, authsource='twitter', readpreference='primary', appname='MongoDB Compass Community', ssl=False), 'ny2019')

In [14]:
ny_db.tweets

Collection(Database(MongoClient(host=['3.14.72.122:27017'], document_class=dict, tz_aware=False, connect=True, authsource='twitter', readpreference='primary', appname='MongoDB Compass Community', ssl=False), 'ny2019'), 'tweets')

In [15]:
databasepopulation.populatetweets(ny_db, path=PATH, cityprefix='ny', yearstart=2019, yearend=2019)

Now populating year  2019
process completed


### ---Tweets populated on `test` database - `tweets` collection!---

## Extra: Experiment with Datetime to Epoch Conversion

In [23]:
ny12.loc[0]['date']

'2012-07-26 00:49:21+00:00'

In [6]:
from datetime import datetime, timezone

In [26]:
from dateutil.parser import parse

# Experiement for old date format
date_str = '2012-07-26 00:49:21+00:00'
data_obj = parse(date_str)
epoch = datetime.utcfromtimestamp(0)
data_obj.replace(tzinfo=timezone.utc).timestamp() * 1000

#expected: 1343263761000

1343263761000.0

Source (Method 2): https://www.techiedelight.com/convert-datetime-object-to-milliseconds-since-epoch-python/#:~:text=Using%20datetime.&text=Starting%20with%20Python%203.3%2C%20you,can%20be%20done%20with%20dt.

In [25]:
# Experiment for new data format
date_str = 'Fri Apr 24 23:56:13 +0000 2020'

date_obj = datetime.strptime(date_str, '%a %b %d %H:%M:%S %z %Y')
date_obj_str = str(date_obj)
epoch = datetime.utcfromtimestamp(0)
date_obj.replace(tzinfo=timezone.utc).timestamp() * 1000


1587772573000.0